File: services/api/src/finai_api/services/company_directory.py
"""Business chooser classification over an already authorized ontology snapshot.

This projection grants no access and changes no identity. Its caller owns RLS and
valid/known-time selection. Retained observations remain available as evidence.
"""

from typing import Any


def project(nodes: list[dict[str, Any]], pins: dict[tuple[str, str], str]) -> dict[str, Any]:
    eligible = {
        node["resource_id"]: node
        for node in nodes
        if node.get("authority_state") == "APPROVED"
        and node.get("evidence_class") in {"USER_ASSERTED", "SOURCE_BOUND"}
    }

    def linked(node: dict[str, Any], field: str, kind: str) -> dict[str, Any] | None:
        target = eligible.get(node["attributes"].get(field))
        if (
            target
            and target["object_type"] == kind
            and pins.get((node["version_id"], field)) == target["version_id"]
        ):
            return target
        return None

    configured: dict[str, list[str]] = {}
    source_ids: set[str] = set()
    reported_ids: set[str] = set()
    for node in eligible.values():
        kind = node["object_type"]
        if kind == "CompanyWorkspace" and node["evidence_class"] == "USER_ASSERTED":
            company = linked(node, "company_id", "LegalEntity")
            if (
                company
                and linked(node, "enterprise_id", "EnterpriseGroup")
                and linked(node, "domain_pack_id", "DomainPack")
            ):
                configured.setdefault(company["resource_id"], []).append(node["resource_id"])
        elif kind == "SourceAccountingScope":
            company = linked(node, "legal_entity_id", "LegalEntity")
            if company:
                source_ids.add(company["resource_id"])
        elif kind == "CorporateDisclosureBinding":
            if linked(node, "observation_id", "SourceCorporateObservation"):
                for field in ("reporter_id", "related_entity_id"):
                    company = linked(node, field, "LegalEntity")
                    if company:
                        reported_ids.add(company["resource_id"])

    companies, sources, reported, unclassified = [], [], [], []
    for node in eligible.values():
        if node["object_type"] != "LegalEntity":
            continue
        identifier = node["resource_id"]
        declared = node["evidence_class"] == "USER_ASSERTED"
        if declared or identifier in configured:
            companies.append(
                {
                    "company": node,
                    "basis": "EXPLICIT_COMPANY_DECLARATION" if declared else "CONFIGURED_WORKSPACE",
                    "workspace_ids": sorted(configured.get(identifier, [])),
                }
            )
            continue
        # These are producer identity namespaces, never name/region heuristics.
        source = identifier in source_ids or node["identity_key"].startswith("observed-company:")
        filing = identifier in reported_ids or node["identity_key"].startswith("reported-code:")
        if source:
            sources.append(node)
        if filing:
            reported.append(node)
        if not source and not filing:
            unclassified.append(node)
    return {
        "contract": "company-directory/1",
        "companies": companies,
        "source_identities": sources,
        "reported_parties": reported,
        "unclassified_identities": unclassified,
    }

File: apps/web/app/company-directory.ts
import type {CanonicalResource} from "@finai/contracts";

export type CompanyDirectory={contract:"company-directory/1";companies:{company:CanonicalResource;basis:"EXPLICIT_COMPANY_DECLARATION"|"CONFIGURED_WORKSPACE";workspace_ids:string[]}[];source_identities:CanonicalResource[];reported_parties:CanonicalResource[];unclassified_identities:CanonicalResource[]};
const uuid=/^[a-f0-9]{8}(?:-[a-f0-9]{4}){3}-[a-f0-9]{12}$/i,hash=/^[a-f0-9]{64}$/;
function node(value:unknown):value is CanonicalResource {if(!value||typeof value!=="object")return false;const v=value as CanonicalResource;return typeof v.resource_id==="string"&&uuid.test(v.resource_id)&&typeof v.version_id==="string"&&uuid.test(v.version_id)&&typeof v.content_hash==="string"&&hash.test(v.content_hash)&&typeof v.display_name==="string"&&typeof v.object_type==="string"&&typeof v.evidence_class==="string"&&Boolean(v.attributes)&&typeof v.attributes==="object"&&!Array.isArray(v.attributes);}
/** Consume the server classification; legacy index collections never grant selection. */
export function companyDirectory(index:unknown):CompanyDirectory|null {
 if(!index||typeof index!=="object")return null;const d=(index as {company_directory?:CompanyDirectory}).company_directory;
 if(!d||d.contract!=="company-directory/1"||!Array.isArray(d.companies)||![d.source_identities,d.reported_parties,d.unclassified_identities].every(rows=>Array.isArray(rows)&&rows.every(node)))return null;
 const seen=new Set<string>();
 for(const entry of d.companies){if(!entry||!node(entry.company)||entry.company.object_type!=="LegalEntity"||entry.company.authority_state!=="APPROVED"||!["USER_ASSERTED","SOURCE_BOUND"].includes(entry.company.evidence_class)||!["EXPLICIT_COMPANY_DECLARATION","CONFIGURED_WORKSPACE"].includes(entry.basis)||!Array.isArray(entry.workspace_ids)||entry.workspace_ids.some(id=>typeof id!=="string"||!uuid.test(id))||new Set(entry.workspace_ids).size!==entry.workspace_ids.length||entry.basis==="CONFIGURED_WORKSPACE"&&!entry.workspace_ids.length||entry.basis==="EXPLICIT_COMPANY_DECLARATION"&&entry.company.evidence_class!=="USER_ASSERTED"||seen.has(entry.company.resource_id))return null;seen.add(entry.company.resource_id);}
 const reviewed=new Set<string>();
 for(const rows of [d.source_identities,d.reported_parties]){const ids=new Set<string>();for(const item of rows){if(seen.has(item.resource_id)||ids.has(item.resource_id))return null;ids.add(item.resource_id);reviewed.add(item.resource_id);}}
 const unclassified=new Set<string>();for(const item of d.unclassified_identities){if(seen.has(item.resource_id)||reviewed.has(item.resource_id)||unclassified.has(item.resource_id))return null;unclassified.add(item.resource_id);}
 return d;
}
export function selectableCompanies(index:unknown):CanonicalResource[]{return companyDirectory(index)?.companies.map(entry=>entry.company)??[];}

export const readCompanyDirectory=companyDirectory;
